zipkin-storage/cassandra3/src/main/java/zipkin/storage/cassandra3/CassandraSpanStore.java

```diff
@@ -0,0 +1,507 @@
+            "annotations", "binary_annotations")
+            .from(TABLE_TRACES)
+            .where(QueryBuilder.in("trace_id", QueryBuilder.bindMarker("trace_id")))
+            .limit(QueryBuilder.bindMarker("limit_")));
+
+    selectDependencies = session.prepare(
+        QueryBuilder.select("links")
+            .from(Schema.TABLE_DEPENDENCIES)
+            .where(QueryBuilder.in("day", QueryBuilder.bindMarker("days"))));
+
+    selectServiceNames = session.prepare(
+        QueryBuilder.select("service_name")
+            .distinct()
+            .from(Schema.VIEW_TRACE_BY_SERVICE));
+
+    selectSpanNames = session.prepare(
+        QueryBuilder.select("span_name")
+            .from(Schema.VIEW_TRACE_BY_SERVICE)
+            .where(QueryBuilder.eq("service_name", QueryBuilder.bindMarker("service_name")))
+            .limit(QueryBuilder.bindMarker("limit_")));
+
+    selectTraceIdsByServiceSpanName = session.prepare(
+        QueryBuilder.select("ts", "trace_id")
+            .from(TABLE_TRACE_BY_SERVICE_SPAN)
+            .where(QueryBuilder.eq("service_name", QueryBuilder.bindMarker("service_name")))
+            .and(QueryBuilder.eq("span_name", QueryBuilder.bindMarker("span_name")))
+            .and(QueryBuilder.eq("bucket", QueryBuilder.bindMarker("bucket")))
+            .and(QueryBuilder.gte("ts", QueryBuilder.bindMarker("start_ts")))
+            .and(QueryBuilder.lte("ts", QueryBuilder.bindMarker("end_ts")))
+            .limit(QueryBuilder.bindMarker("limit_")));
+
+    Select durationQuery = QueryBuilder.select("ts", "trace_id")
+        .from(TABLE_TRACE_BY_SERVICE_SPAN)
+        .where(QueryBuilder.eq("service_name", QueryBuilder.bindMarker("service_name")))
+        .and(QueryBuilder.eq("span_name", QueryBuilder.bindMarker("span_name")))
+        .and(QueryBuilder.eq("bucket", QueryBuilder.bindMarker("bucket")))
+        .and(QueryBuilder.gte("ts", QueryBuilder.bindMarker("start_ts")))
+        .and(QueryBuilder.lte("ts", QueryBuilder.bindMarker("end_ts")))
+        .and(QueryBuilder.gte("duration", QueryBuilder.bindMarker("start_duration")))
+        .and(QueryBuilder.lte("duration", QueryBuilder.bindMarker("end_duration")))
+        .limit(QueryBuilder.bindMarker("limit_"));
+
+    if (Schema.readMetadata(session).allowFilteringOnDuration) {
+      durationQuery.allowFiltering();
+    }
+
+    selectTraceIdsByServiceSpanNameAndDuration = session.prepare(durationQuery);
+
+    selectTraceIdsByAnnotation = session.prepare(
+        QueryBuilder.select("ts", "trace_id")
+            .from(TABLE_TRACES)
+            .where(QueryBuilder.contains("annotation_keys", QueryBuilder.bindMarker("annotation_key")))
+            .and(QueryBuilder.gte("ts_uuid", QueryBuilder.bindMarker("start_ts")))
+            .and(QueryBuilder.lte("ts_uuid", QueryBuilder.bindMarker("end_ts")))
+            .limit(QueryBuilder.bindMarker("limit_"))
+            .allowFiltering());
```

Comment:
this won't work in any real installation because it is doing a full table scan every request.

this approach (using set<string> for annotations in the trace table) would have to use a separate new materialized view. which then becomes a different approach to SASI for annotations as I had originally proposed. my original proposal was the fast approach, and simpler in the long run.